[MRG+1] Traceback message in Gallery output

Titan-C opened the pull request:
This PR need comments. 
First there is the situation of how to deal with a broken build #45, discussion shall continue there.
Here I address, for the moment, that the traceback of a broken example should be also printed in the HTML output of the gallery following my comment on https://github.com/sphinx-gallery/sphinx-gallery/issues/65#issuecomment-151472930

agramfort commented:
that's fine with me

other option is to stop the doc build with a meaningful crash message

GaelVaroquaux commented:
> other option is to stop the doc build with a meaningful crash message

It should be controlled by an option (just like the building of the
gallery is).

lesteve commented:
>> other option is to stop the doc build with a meaningful crash message

> It should be controlled by an option (just like the building of the gallery is).

My 2 cents: I think the current behaviour is fine for now. The general use case is to build all the examples to find all the broken ones rather than fail on the first error.

Something I would actually find useful (not sure how easy this is): have a special image for broken examples allowing them to identify quite easily in the gallery.

lesteve commented on sphinx_gallery/gen_rst.py:
```diff
@@ -447,12 +447,14 @@ def execute_script(code_block, example_globals, image_path, fig_count,
         print(80 * '_')
+
+        figure_list = []
+        exception_msg = indent(traceback.format_exc(), '    ')
```
Nitpick: in order not to do traceback.print_exc() and then traceback.format_exc() which is kind of doing the same thing twice, you can do:

```python

formatted_exception = traceback.format_exc()

print(formatted_exception)

exception_msg = indent(formatted_exception, ' ' * 4)
```

lesteve commented:
Just tried this out and here is a snapshot:

![sphinx_gallery_traceback](https://cloud.githubusercontent.com/assets/1680079/11867691/f97ef18e-a4b2-11e5-89ad-2855b74be4fc.png)

It would be nice to have a fixed-width font in the HTML traceback. This makes a huge difference for long tracebacks.

Titan-C commented:
Now it's improved.
I painted the broken image that takes over in the gallery thumbnails on failed examples. And the traceback message is not anymore in the sphinx warning box that changes font style but uses just a container.

lesteve commented:
> Now it's improved.
I painted the broken image that takes over in the gallery thumbnails on failed examples. And the traceback message is not anymore in the sphinx warning box that changes font style but uses just a container.

The image thing is quite nice, although we can think about tweaking the "broken" image.

I still get non fixed-width font in the warnings when generating the gallery locally though.

Titan-C commented:
> I still get non fixed-width font in the warnings when generating the gallery locally though.

No idea here. What does you browser tell you the CSS style is. The block shall be a sphx-glr-traceback. That one has no style for fonts. The example I'm providing gives me this

![snapshot1](https://cloud.githubusercontent.com/assets/713451/11872995/a08e696c-a4d8-11e5-9325-f506e95609f8.png)

lesteve commented:
I have to say not sure whether the technical term for it is fixed-font or not but basically what I was trying to say is that I would have the same font in the traceback and in the code cell.

Titan-C commented:
> I would have the same font in the traceback and in the code cell.

I just learned that is syntax hightlighting for python traceback messages. Thus I'm making the traceback outputs be parsed as that. Plus an additional light red box to call attention is a traceback and that the example fails.

![image](https://cloud.githubusercontent.com/assets/713451/12034668/9dd95be0-ae34-11e5-9d34-414ca25e83df.png)

Titan-C commented:
As here I change some CSS I would like PR #80 to be merged first. This branch is already rebased to it, but it did not generate a branch dependency as I expected.

agramfort commented:
is it possible to have the build stop if an example fails? to me that should be the default behavior. It happened to me too many times to push a website update with broken examples... it happens quickly when you have dozens of examples.

Titan-C commented:
I added the quick fail option to break the gallery build process as soon as an example trows an Exception, added the documentation what to put in the makefile for that. But I have not made it the default behavior, as building the complete gallery and scanning for the broken examples visually from the rendered website seems more intuitive for me. You can vote on that.

agramfort commented on doc/advanced_configuration.rst:
```diff
@@ -163,3 +163,50 @@ This will download directly from PyPI our latest released code and
+
+Sphinx-Gallery can parse all your examples and build the gallery
+without executing any of the scripts. This is just for speed
+visiulization processes of the gallery and the size it take your
```
visiulization

agramfort replied:
take -> takes

display -> to display

agramfort commented on doc/advanced_configuration.rst:
```diff
@@ -163,3 +163,50 @@ This will download directly from PyPI our latest released code and
+Sphinx-Gallery can parse all your examples and build the gallery
+without executing any of the scripts. This is just for speed
+visiulization processes of the gallery and the size it take your
+website display, or any use you can imagine for it. To achive this you
```
achive -> achieve

agramfort commented on doc/advanced_configuration.rst:
```diff
@@ -163,3 +163,50 @@ This will download directly from PyPI our latest released code and
+
+As your project evolves some of your example scripts might stop
+executing properly. Sphinx-Gallery assist you in the discovery process
+of those bugged examples. The defaul behavior is to replace the
```
defaul -> default

agramfort commented on doc/advanced_configuration.rst:
```diff
@@ -163,3 +163,50 @@ This will download directly from PyPI our latest released code and
+by including in your ``Makefile``::
+
+    html-quick_fail:
+        $(SPHINXBUILD) -D quick_fail=1 -b html $(ALLSPHINXOPTS) $(BUILDDIR)/html
```
quick_fail -> fail_on_error ?

Titan-C replied:
This one I was also thinking, early_fail, early_abort

agramfort replied:
> This one I was also thinking, early_fail, early_abort

I find fail_on_error more explicit.

agramfort commented:
besides LGTM

GaelVaroquaux commented on sphinx_gallery/gen_rst.py:
```diff
@@ -438,21 +438,35 @@ def execute_script(code_block, example_globals, image_path, fig_count,
         print(80 * '_')
+
+        figure_list = []
+        image_list = codestr2rst(formatted_exception, 'pytb')
```
For readability, I would suggest being explicit about the keyword argument here, by writing:

<pre>
    ..., lang='pytb')
</pre>

GaelVaroquaux commented:
I agree with your various design choices (in particular the default).

+1 for merge. I have suggested a tiny style change.

agramfort commented:
it's fine to merge as is for me

thx @Titan-C

lesteve commented on doc/Makefile:
```diff
@@ -58,6 +58,11 @@ html-noplot:
 	@echo
 	@echo "Build finished. The HTML pages are in $(BUILDDIR)/html."
 
+html-quick_fail:
```
maybe html-quick-fail in order not to mix - and _

lesteve commented on examples/plot_raise.py:
```diff
@@ -0,0 +1,33 @@
+###############################################################################
+# Here is another error raising Block
+
+raise IOError
```
Maybe put a more realistic error like: `plt.plot('this is not going to work')`

lesteve replied:
What I meant is that the traceback in this case will be more interesting.

lesteve commented:
Is broken_stamp.svg actually used?

Slightly related is do you know why the "broken" image png appears with a black background in the gallery:
![sphinx-gallery](https://cloud.githubusercontent.com/assets/1680079/12089835/51d09aa6-b2e9-11e5-9e75-ab834f1585a3.png)

lesteve commented:
My contribution for the 'quick_fail' naming: abort_on_example_error.

lesteve commented:
Could you rebase on master so that we make sure no extra CSS changes has slipped through in this PR?

Titan-C commented:
broken_stamp.svg is not used now, but is the source of for the png file. Not sure if it will become useful in the future if we want to tune the image or if #61 

The background of the image is black in the thumbnails because in the sphinx-classic css theme all body backgrounds are black, and the png image is transparent.

lesteve commented on examples/plot_raise.py:
```diff
@@ -0,0 +1,33 @@
+# -*- coding: utf-8 -*-
+"""
+=============================
+Examples that fail to execute
```
Singular seems more fitting here: Example that fails to execute

GaelVaroquaux commented:
Overall, LGTM. +1 for merge (ideally, please address the comments of the others).
Thanks!

agramfort commented:
fine with me

Titan-C commented:
Rebased to master and comments addressed

GaelVaroquaux commented:
@agramfort was good with this. I am too. Let's merge!
